src/platform/intel/cavs/platform.c

```diff
@@ -391,11 +401,17 @@ int platform_init(struct sof *sof)
 #elif CONFIG_CANNONLAKE
 
 	/* initialize PM for boot */
+
+	/* request configured ring oscillator and wait for status ready */
+	shim_write(SHIM_CLKCTL, shim_read(SHIM_CLKCTL) | CNL_DEFAULT_RO);
+	while (!(shim_read(SHIM_CLKSTS) & CNL_DEFAULT_RO))
+		idelay(16);
```

Comment:
We should bail out here and assert or trace_error() if the clock does not change within a timeout. This will help integrators debugging hangs.

Reply:
There is one. If the fw is waiting too long in this loop, you'd get:

[    5.395706] sof-audio-pci 0000:00:1f.3: error: firmware boot failure
[    5.395723] sof-audio-pci 0000:00:1f.3: status: fw entered - code 00000005
[    5.395736] sof-audio-pci 0000:00:1f.3: error: status = 0x00000000 panic = 0x00004190

in the dmesg where 0x4190 code is exactly `TRACE_BOOT_PLATFORM_CPU_FREQ`.

Reply:
Sorry, I mean there is no timeout in this loop, it can potentially loop forever.

Reply:
If the fw exits the loop after t/o and panics with a new error code, how would that be different from the information currently reported (example pasted above) to an integrator in case of hang?

Reply:
Sorry not following.
```c
while (!(shim_read(SHIM_CLKSTS) & CNL_DEFAULT_RO))
		idelay(16);
```
How would this exit if the HW condition is never satisfied ?

Reply:
Hypothetically, assumed t/o is added in this place -is there any additional value? The entire initialization sequence is already protected by the driver, that reports the last trace point if t/o occurs. I provided example of info that is reported to the dmesg if the dsp keeps polling the clock status too long. This is how the existing initialization trace point mechanism + t/o implemented in the driver provides imo sufficient information about initialization progress and potential issues. That is why I asked what would be the real gain of implementing another t/o mechanism for individual micro steps in initialization phase?

Reply:
In general it's good practice to have t/o logic in all while(hw_state) loops. This way we can assert() with the line number etc. The trace_point is good, but it does mean more digging and is not guranteed to have a 1:1 mapping with failures (as many steps can be taken in between different trace points).

Reply:
We can change these sets to what's going to be introduced in https://github.com/thesofproject/sof/pull/1968

Anyway there are no timeouts either and I think they are not needed as driver will t/o, so it's just unnecessary redundancy.

Reply:
@lgirdwood I fully agree, the t/o logic is a must-have in the *run-time* code. However this piece of *initialization* code is already nicely protected by the driver will immediate information about the phase that failed. So I'd rather keep the trace points mapping precise (which already is in this particular case). Esp. that you may have some blocking unprotectable reads, so having a single common t/o mechanism, maintained to be consistent is a better option in this case imo.